### ISSUE TITLE
User should not have to provide an "id" when adding an API definition

### ISSUE THREAD

**Issue by marians** [TykTechnologies/tyk#943]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

When submitting a new API via `POST /tyk/apis/` without setting the `"id"` attribute, the request is accepted and a new API definition JSON is created in the `apps` directory. But the new API is not proxied.

In addition, the ID provided has to meet the criteria of a [BSON ObjectID](https://docs.mongodb.com/manual/reference/method/ObjectId/). Just providing a unique string won't work. This fact is only visible in the [source](https://github.com/TykTechnologies/tyk/blob/master/apidef/api_definitions.go#L262).

**What is the expected behavior?**

If an ID is required, tyk should create one if it's missing.

If an ID is submitted that doesn't meet the criteria of a BSON ObjectId, tyk should give specific feedback in logs and in the API response.

As the [docs](https://tyk.io/docs/tyk-rest-api/api-definition-object-details/) state, "This is only used in the MongoDB context and bears no actual relation to the identity of the API." - As I am using tyk without Dashboard and without MongoDB, the need for this ID as well as the format restrictions come totally unexpected.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**

Using [httpie](https://httpie.org/) here for convenience.

1. Start a tyk without any API definitions in `apps`, listening to `localhost:8080`.

2. Submit an API definition via the tyk API

```nohighlight
# our Tyk API secret as configured
export TYK_SECRET=352d20ee67be67f6340b4c0605b044bc4
export AUTH_HEADER="x-tyk-authorization:${TYK_SECRET}"

# submit a minimal API definition, no "id"
http post localhost:8080/tyk/apis/ $AUTH_HEADER \
  api_id=sparse-api \
  proxy:='{"listen_path": "/sparse-api/", "target_url": "http://httpbin.org/headers", "strip_listen_path": true}'
```

3. Inspect the created file `apps/sparse-api.json`. There should not be an  `"id"` attribute in it.

4. Reload config, just to be sure.

```nohighlight
http get localhost:8080/tyk/reload/ $AUTH_HEADER
HTTP/1.1 200 OK
...
```

5. Check whether the API is proxied.

```nohighlight
http get localhost:8080/sparse-api/
HTTP/1.1 404 Not Found
...
```

6. Edit `apps/sparse-api.json` to add an ID like `"id": "55780c119b23c3000200004f"`.

7. Reload again

6. Check again whether the API is proxied.

It should now get proxied.

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**

Tested in 2.3.7. Don't know about earlier versions.

**Comment by buger** [comment]
Thank you! This indeed looks like a bug. Community version should not require id in such format.

**Comment by mvdan** [comment]
This is surprising - our unit tests never set this bson ID.

**Comment by mvdan** [comment]
I just tried to reproduce this on both 2.3.7 and master following these steps:

* Write `apps/app_sample.json` (content below)
* Start gateway
* Open browser to `http://localhost:8080/test/` (it works)

API Definition:

```
{
	"api_id": "1",
	"org_id": "default",
	"definition": {
		"location": "header",
		"key": "version"
	},
	"use_keyless": true,
	"version_data": {
		"not_versioned": true,
		"versions": {"Default": {"name": "Default"}}
	},
	"proxy": {
		"listen_path": "/test/",
		"target_url": "http://httpbin.org",
		"strip_listen_path": true
	}
}
```

@marians could you please check if restarting the gateway is any better than issuing a reload? What about writing the file to disk and restarting the gateway, instead of inserting the API definition via the gateway's own API?

**Comment by mvdan** [comment]
I just tried again, this time even installing httpie and strictly following the instructions.

This was on 2.3.7, with the default config as written by `tyk --conf=tyk.conf` when that file doesn't exist, and with `apps/` empty.

Note that you do get a `Version information not found`, but that's a 403, not a 404.

Also note that reloads aren't blocking, and the gateway limits the amount of reloads - as of 2.3.7, that's one reload every ten seconds. If multiple reloads are issued, a single one gets queued and the rest are ignored.

So have you tried waiting for 10 seconds after you issue the reload? I too got a 404 in a script that immediately pinged the API after issuing a reload. But once the script was finished and I pinged it again a few seconds later, it was already up.

See https://github.com/TykTechnologies/tyk/pull/717 too - that could be helpful. Also note that the upcoming release, 2.4, lowers the throttling of reloads from 10s to 1s.
